File: scratch/analyze_lint.py
import json
import os

path = 'scratch/lint_report.json' if os.path.exists('scratch/lint_report.json') else 'eslint_report.json'

with open(path, 'r', encoding='utf-8', errors='ignore') as f:
    text = f.read()

# find first '[' and last ']'
start = text.find('[')
end = text.rfind(']')

if start != -1 and end != -1:
    json_str = text[start:end+1]
    data = json.loads(json_str)

    rule_counts = {}
    total_errors = 0
    total_warnings = 0
    file_errors = {}
    file_warnings = {}

    for file in data:
        rel_path = os.path.relpath(file['filePath'], '.')
        err_c = 0
        warn_c = 0
        for msg in file.get('messages', []):
            rule = msg.get('ruleId') or 'syntax/other'
            if rule not in rule_counts:
                rule_counts[rule] = {'error': 0, 'warning': 0, 'sample_msg': msg.get('message', '')}
            
            if msg.get('severity') == 2:
                rule_counts[rule]['error'] += 1
                total_errors += 1
                err_c += 1
            else:
                rule_counts[rule]['warning'] += 1
                total_warnings += 1
                warn_c += 1
        if err_c > 0:
            file_errors[rel_path] = err_c
        if warn_c > 0:
            file_warnings[rel_path] = warn_c

    print(f"Total Errors: {total_errors}")
    print(f"Total Warnings: {total_warnings}")
    print("\n--- BREAKDOWN BY RULE ---")
    for r, c in sorted(rule_counts.items(), key=lambda x: x[1]['error'] + x[1]['warning'], reverse=True):
        print(f"Rule: {r:<40} | Errors: {c['error']:<5} | Warnings: {c['warning']:<5} | Sample: {c['sample_msg'][:60]}")

    print("\n--- TOP FILES WITH ERRORS ---")
    for f, c in sorted(file_errors.items(), key=lambda x: x[1], reverse=True)[:15]:
        print(f"{f}: {c} errors")
else:
    print("Could not locate JSON array in report.")

File: scratch/analyze_lint.js
const fs = require('fs');
const path = require('path');

const reportPath = fs.existsSync('./scratch/lint_report.json') ? './scratch/lint_report.json' : './eslint_report.json';
const text = fs.readFileSync(reportPath, 'utf8');

const start = text.indexOf('[');
const end = text.lastIndexOf(']');

if (start !== -1 && end !== -1) {
  const jsonStr = text.substring(start, end + 1);
  const data = JSON.parse(jsonStr);

  const ruleCounts = {};
  let totalErrors = 0;
  let totalWarnings = 0;
  const fileErrors = {};
  const fileWarnings = {};

  data.forEach(file => {
    const relPath = path.relative('.', file.filePath);
    let errC = 0;
    let warnC = 0;
    (file.messages || []).forEach(msg => {
      const rule = msg.ruleId || 'syntax/other';
      if (!ruleCounts[rule]) {
        ruleCounts[rule] = { error: 0, warning: 0, sampleMsg: msg.message };
      }
      if (msg.severity === 2) {
        ruleCounts[rule].error++;
        totalErrors++;
        errC++;
      } else {
        ruleCounts[rule].warning++;
        totalWarnings++;
        warnC++;
      }
    });
    if (errC > 0) fileErrors[relPath] = errC;
    if (warnC > 0) fileWarnings[relPath] = warnC;
  });

  console.log(`Total Errors: ${totalErrors}`);
  console.log(`Total Warnings: ${totalWarnings}`);
  console.log('\n--- BREAKDOWN BY RULE ---');
  Object.entries(ruleCounts)
    .sort((a, b) => (b[1].error + b[1].warning) - (a[1].error + a[1].warning))
    .forEach(([rule, counts]) => {
      console.log(`Rule: ${rule.padEnd(42)} | Errors: ${String(counts.error).padEnd(5)} | Warnings: ${String(counts.warning).padEnd(5)} | Sample: ${counts.sampleMsg.substring(0, 60)}`);
    });

  console.log('\n--- TOP FILES WITH ERRORS ---');
  Object.entries(fileErrors)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 15)
    .forEach(([f, c]) => console.log(`${f}: ${c} errors`));
} else {
  console.log('Could not locate JSON array in report.');
}
